segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3Storage.java

```diff
@@ -0,0 +1,380 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.impl.extendeds3;
+
+import com.emc.object.Range;
+import com.emc.object.s3.S3Client;
+import com.emc.object.s3.S3Exception;
+import com.emc.object.s3.S3ObjectMetadata;
+import com.emc.object.s3.bean.AccessControlList;
+import com.emc.object.s3.bean.CanonicalUser;
+import com.emc.object.s3.bean.CopyPartResult;
+import com.emc.object.s3.bean.Grant;
+import com.emc.object.s3.bean.ListObjectsResult;
+import com.emc.object.s3.bean.MultipartPartETag;
+import com.emc.object.s3.bean.Permission;
+import com.emc.object.s3.request.CompleteMultipartUploadRequest;
+import com.emc.object.s3.request.CopyPartRequest;
+import com.emc.object.s3.request.PutObjectRequest;
+import com.emc.object.s3.request.SetObjectAclRequest;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import io.pravega.common.LoggerHelpers;
+import io.pravega.common.io.StreamHelpers;
+import io.pravega.common.util.ImmutableDate;
+import io.pravega.segmentstore.contracts.BadOffsetException;
+import io.pravega.segmentstore.contracts.SegmentProperties;
+import io.pravega.segmentstore.contracts.StreamSegmentExistsException;
+import io.pravega.segmentstore.contracts.StreamSegmentInformation;
+import io.pravega.segmentstore.contracts.StreamSegmentNotExistsException;
+import io.pravega.segmentstore.contracts.StreamSegmentSealedException;
+import io.pravega.segmentstore.storage.SegmentHandle;
+import io.pravega.segmentstore.storage.impl.filesystem.IdempotentStorageBase;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.AccessDeniedException;
+import java.time.Duration;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Storage adapter for extended S3 based storage.
+ *
+ * Each segment is represented as a single Object on the underlying storage.
+ *
+ * This implementation works under the assumption that data is only appended and never modified.
+ * Each block of data has an offset assigned to it and Pravega always writes the same data to the same offset. As a result
+ * the only flow when a write call is made to the same offset twice is when ownership of the segment changes
+ * from one host to another and both the hosts are writing to it.
+ *
+ * As PutObject calls to with the same start-offset to an Extended S3 object are idempotent (any attempt to re-write data with the same file offset does not
+ * cause any form of inconsistency), fencing is not required.
+ *
+ * Here is the expected behavior in case of ownership change: both the hosts will keep writing the same data at the same offset till the time the
```

Comment:
This is not enough detail, could you add the actual path for the ownership change? For example, start by referring to `ZKSegmentContainerMonitor`...